rust/lance-encoding/src/data.rs

```diff
@@ -372,17 +307,851 @@ impl DataBlock for DictionaryDataBlock {
+// index type.  If we do, we need to upscale the indices to a larger type.
+fn concat_dict_arrays(arrays: &[ArrayRef]) -> ArrayRef {
+    let value_type = arrays[0].as_any_dictionary().values().data_type();
+    let array_refs = arrays.iter().map(|arr| arr.as_ref()).collect::<Vec<_>>();
+    match arrow_select::concat::concat(&array_refs) {
+        Ok(array) => array,
+        Err(arrow_schema::ArrowError::DictionaryKeyOverflowError { .. }) => {
+            // Slow, but hopefully a corner case.  Optimize later
+            let upscaled = array_refs
+                .iter()
+                .map(|arr| {
+                    match arrow_cast::cast(
+                        *arr,
+                        &DataType::Dictionary(
+                            Box::new(DataType::UInt32),
+                            Box::new(value_type.clone()),
+                        ),
+                    ) {
+                        Ok(arr) => arr,
+                        Err(arrow_schema::ArrowError::DictionaryKeyOverflowError { .. }) => {
+                            // Technically I think this means the input type was u64 already
+                            unimplemented!("Dictionary arrays with more than 2^32 unique values")
+                        }
+                        err => err.unwrap(),
+                    }
+                })
+                .collect::<Vec<_>>();
+            let array_refs = upscaled.iter().map(|arr| arr.as_ref()).collect::<Vec<_>>();
+            // Can still fail if concat pushes over u32 boundary
+            match arrow_select::concat::concat(&array_refs) {
+                Ok(array) => array,
+                Err(arrow_schema::ArrowError::DictionaryKeyOverflowError { .. }) => {
+                    unimplemented!("Dictionary arrays with more than 2^32 unique values")
+                }
+                err => err.unwrap(),
+            }
+        }
+        // Shouldn't be any other possible errors in concat
+        err => err.unwrap(),
+    }
+}
+
+fn max_index_val(index_type: &DataType) -> u64 {
+    match index_type {
+        DataType::Int8 => i8::MAX as u64,
+        DataType::Int16 => i16::MAX as u64,
+        DataType::Int32 => i32::MAX as u64,
+        DataType::Int64 => i64::MAX as u64,
+        DataType::UInt8 => u8::MAX as u64,
+        DataType::UInt16 => u16::MAX as u64,
+        DataType::UInt32 => u32::MAX as u64,
+        DataType::UInt64 => u64::MAX,
+        _ => panic!("Invalid dictionary index type"),
+    }
+}
+
+// If we get multiple dictionary arrays and they don't all have the same dictionary
+// then we need to normalize the indices.  Otherwise we might have something like:
+//
+// First chunk ["hello", "foo"], [0, 0, 1, 1, 1]
+// Second chunk ["bar", "world"], [0, 1, 0, 1, 1]
+//
+// If we simply encode as ["hello", "foo", "bar", "world"], [0, 0, 1, 1, 1, 0, 1, 0, 1, 1]
+// then we will get the wrong answer.
+//
+// A simple way to do this today is to just concatenate all the arrays.
```

Comment:
could you elaborate a bit on this?

Reply:
I read the [arrow-rs's concat_dictionaries](https://docs.rs/arrow-select/52.2.0/src/arrow_select/concat.rs.html#85) and get a understanding now

Reply:
I expanded on this comment a little.